### ISSUE TITLE
Unify `let` and `where`

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#6149]
Lunch discussion 2022-09-30: We could interpret `Gamma |- ... let defs in e` as
```agda
module M Gamma where
  defs

Gamma |- ... open M Gamma in e
```
This would unify the `let` and `where` constructs: we could allow arbitrary `defs` in a `let` then.

1. Would this cause regressions, i.e., programs that do not type-check under the new semantics?
2. Currently `let` allow record pattern matches as in `let (x , y) = foo in e`, and `where` doesn't.  A first step would be to also allow these in `where` blocks.
3. Does this new semantics cause problems when forming the record constructor?

**Comment by jespercockx** [comment]
Would this mean that all variables in a let-binding get lambda-lifted? In that case I do not think this is a good idea.

**Comment by sattlerc** [comment]
In the long term, the parametrized module semantics should change to eliminate lambda lifting.

**Comment by jespercockx** [comment]
Yes, but then that should be done *before* we unify `let` and `where`.

**Comment by sattlerc** [comment]
Related: #2795

**Comment by plt-amy** [comment]
> Currently let allow record pattern matches as in let (x , y) = foo in e, and where doesn't. A first step would be to also allow these in where blocks.

Would we allow these at the top-level, too? Edit: d'oh, I should've clicked through to 2795.

**Comment by uhbif19** [comment]
> In the long term, the parametrized module semantics should change to eliminate lambda lifting.

Why so? Or other way around - why it was needed before?

**Comment by jespercockx** [comment]
- https://github.com/agda/agda/issues/426

**Comment by uhbif19** [comment]
Yes, I found this issue starting from #426 :D So my question is why lambda lifting was needed in first place. By searching code I did not found relevant, while I did not try to research on parametric modules.

**Comment by jespercockx** [comment]
Well how else would you handle parametrized modules? There is no such thing in Agda's internal syntax, so you need to desugar it to *something*.

**Comment by JacquesCarette** [comment]
I thought it was now 'conventional wisdom' that there should be something like `let` in Agda's internal syntax, but no one has done it because it is a large job. Is that an incorrect impression?

**Comment by jespercockx** [comment]
I mean, yes, I was just answering the question "why was lambda lifting needed in the first place". Having some kind of let bindings and/or closures in internal syntax is indeed the right way to fix this properly.

**Comment by uhbif19** [comment]
*...stuff removed*

Found two mentions of recent talks on module system, but did not found their content.

**UPD**

Found thesis by Ivar Cornelis de Bruin, it is awesome, and will probably cover most of my questions.

Parameterized module de-sugaring probably happens here:
https://github.com/agda/agda/blob/3d393d65baefa07b70f4d3e55475b7852cbe12e5/src/full/Agda/TypeChecking/Rules/Decl.hs#L937

**Comment by uhbif19** [comment]
@jespercockx @sattlerc 

> In the long term, the parametrized module semantics should change to eliminate lambda lifting.

> Yes, but then that should be done before we unify let and where.

Code looks like parametrized modules application are compiled to `LetApply` construct, while regular `let` bindings do not. So could not this be used to make regular lets be working like wheres, while parameterized module lets remain to work as they do now?
